### PR TITLE
feat: serve agents, teams, and workflows as MCP tools

### PR DESCRIPTION
**Pull request by ashpreetbedi**
## Summary

Serve AgentOS components directly as MCP tools, plus the MCP API rename set.

**Components as tools — `tools=` is the one surface.** `MCPConfig.tools` accepts plain callables (custom tools, unchanged), bare `Agent`/`Team`/`Workflow` instances (tool named after the id, described by the component), and `component.as_tool(name=..., description=...)` — a declarative marker publishing the component under a model-facing name and pitch decoupled from it (a tool description is a prompt for the *calling* model; the component's description is written for humans):

```python
agent_os = AgentOS(
    agents=[chief, researcher],
    mcp=MCPConfig(
        default_tools=False,
        tools=[
            chief,
            researcher.as_tool(
                name="deep_research",
                description="Thorough, sourced research. Send one clear question.",
            ),
        ],
    ),
)
# tools/list -> ["chief", "deep_research", "continue_run", "cancel_run"]
```

**HITL works out of the box (local components):** `continue_run`/`cancel_run` are dual-tagged `{core, lifecycle}` and the lifecycle pair rides along whenever components are exposed — even with `default_tools=False` — so a run that pauses on a confirmation-required tool is resumable over MCP (the paused result's `structuredContent` carries the component id, run_id, session_id, and requirements that `continue_run` needs). The pair is bounded to the publication list -- when it rides along (rather than being served as core or explicitly included), it refuses runs of unpublished roster components -- and is scope-gated per component. Opt out with `lifecycle_tools=False` (paused runs then point at REST) or an explicit `exclude_tags={"lifecycle"}`; `include_tags={"lifecycle"}` is also expressible on its own. The full pause → continue loop is covered by a test driving an `as_tool`-renamed exposure end to end.

The generated tools introduce **no new run semantics — only new names**: each is a closure over the component's id that resolves at call time through the exact `run_agent`/`run_team`/`run_workflow` chain (`_require_tool_scopes` → `_resolve_user_id` → `_resolve_run_component` with per-run copies/registry/factories → fresh-session minting → session-ownership gate → streaming progress → `build_run_tool_result`). RBAC scopes (`agents:run`, per-resource variants) apply identically.

Guard rails:
- Exposed components must be in the AgentOS roster (pass the instance; strings are a TypeError) — exposure is a view on the deployment, not a second registration path. Kind derives from roster membership (`Remote*` aren't Agent/Team subclasses), and an equal-id copy matching more than one roster is an ambiguity error instead of a silent wrong-kind publish.
- Tool names (a bare component's id, or the `as_tool` override) are **validated verbatim, never sanitized** — provider shape: leading letter/underscore, `[A-Za-z0-9_-]`, ≤128. The component id — carried in the result's `structuredContent` — remains the `continue_run`/`get_sessions` handle and the per-resource scope segment regardless of the tool name. Collisions with the enabled default tools, custom tools (their *real* FastMCP-registered names), or other exposures are hard build errors — never silent suffixing.
- The tool list is static and identical for every caller — exposure is a deployer-curated publication list (the MCP spec permits per-caller lists; filtering is possible future work); access is denied at call time via scopes.
- The zero-tools validator accepts exposure as a tool source; docs call out the HITL caveat (`continue_run` lives in the default tools).

**API rename set** — all silent aliases, no runtime warnings, conflicting old+new values raise, removal targeted for 3.1:

| Old | New |
|---|---|
| `AgentOS(mcp_server=...)` | `AgentOS(mcp=...)` (old kwarg + property kept as delegating aliases) |
| `MCPServerConfig` | `MCPConfig` (plain-assignment alias export, `isinstance`/imports unchanged) |
| `enable_builtin_tools` | `default_tools` (before-validator alias + read property) |

`mcp_auth` keeps its name.

## Type of change

- [x] New feature
- [x] Improvement

---

## Checklist

- [x] Code complies with style guidelines
- [x] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [x] Self-review completed (five-lens adversarial review workflow over the diff; confirmed findings fixed in the follow-up commit)
- [x] Documentation updated (comments, docstrings)
- [x] Examples and guides: Relevant cookbook examples have been included or updated (if applicable)
- [x] Tested in clean environment
- [x] Tests added/updated (if applicable)

### Duplicate and AI-Generated PR Check

- [x] I have searched existing [open pull requests](../../pulls) and confirmed that no other PR already addresses this issue
- [ ] If a similar PR exists, I have explained below why this PR is a better approach
- [x] Check if this PR was entirely AI-generated (by Copilot, Claude Code, Cursor, etc.)

---

## Additional Notes

**Testing**
- New suite `libs/agno/tests/unit/os/test_mcp_exposed_components.py` (119 tests): exposure surface (names, descriptions, schema, id validation with suggested candidates), session/identity contract (fresh-session minting, JWT-over-caller user_id, scope deny/allow, PAUSED visibility), build-time validation (collisions with default/custom/exposed names, roster membership, zero-tools), and every rename alias including the conflict cases, post-construction `enable_builtin_tools` assignment, and a no-input-mutation check on the pydantic before-validator. Mutation-style guarantees are pinned: call-time resolution (a build-time component capture fails the test), the ownership gate with the right `SessionType` per kind, the workflow factory's own scope gate and session minting, per-resource scopes (`agents:<id>:run`), annotations, the 128-char name cap (id and as_tool paths), and exact description fallback text. Naming guarantees are pinned end to end: provider-shape validation (leading digit, 129-char cap, trailing newline), suggestion quality (collapse, NFKD transliteration, collision-omission), and the auto-derived-id origin note.
- A sync test pins `_BUILTIN_TOOL_NAMES` (the collision map) to what a default server actually registers.
- Full `tests/unit/os` directory at head: 2864 passed, 0 regressions (the telegram files are excluded locally for a pre-existing missing `telebot` dep, untouched by this diff; CI runs them).
- The pre-rename suites (~117 `mcp_server=`/`MCPServerConfig`/`enable_builtin_tools` usages) run **unmodified** against the aliases — the proof the rename is genuinely silent. The only existing-test edits are two message-text pins that deliberately changed spelling with the rename (a `pytest.raises(match=...)` in `test_mcp_oauth.py`, a substring assert in agnoctl's `test_connect_cmd.py`); the asserted behavior is identical in both.
- Live cookbook proof (`cookbook/05_agent_os/14_mcp/agents_as_tools.py`, logged in TEST_LOG.md): real FastMCP client against the served app — `tools/list` returns `chief`/`deep_research` plus the riding `continue_run`/`cancel_run` pair, the sessionless call mints a session and completes, the follow-up on the returned `session_id` recalls the previous answer (re-run live 2026-08-28 at the current head).

**Review fixes (internal re-review of the reshape + lifecycle commits)**
- `exclude_tags` semantics restored: the `lifecycle` tag never enters the enabled set implicitly, so `exclude_tags={"core"}` removes `continue_run`/`cancel_run` exactly as on main (the dual-tag expansion had silently re-added run mutation to read-only surfaces). Upgrade pins cover the pre-lifecycle spellings; an executed enumeration of the full config product showed zero diffs vs main outside the intended ride-along additions.
- Wrong-kind guard restored and extended: entries whose class names their kind (concrete `Agent`/`Team`/`Workflow`, `Remote*`, factories) restrict the id fallback to their own kind's roster and error when a same-id roster component is of a different kind; truly kindless duck-typed entries keep the ambiguity error.
- Collision semantics unified: a custom tool named like an enabled default tool, or like an earlier custom tool, is a hard error (FastMCP would otherwise warn-and-replace, silently shadowing the riding builtin or swallowing the first custom tool). Scoped-out default-tool names remain claimable.
- Roster component factories passed bare in `tools=` classify as exposures (a `BaseFactory` is neither callable nor `arun`-shaped and previously died in `Tool.from_function`).
- Remote components gained `as_tool()`; `model_copy(update={"enable_builtin_tools": ...})` maps the legacy key again; collision errors attribute `as_tool` overrides to the string the user typed; stale docstrings teaching the removed `agents=`/`teams=`/`workflows=` API were rewritten, and the resync staleness contract is documented precisely (post-boot components appear as named tools after restart; under `default_tools=False` no MCP tool can start a run on them until then, while the riding pair still reaches existing runs).
- The riding pair is bounded to the publication list: an executed review escalation showed an exposure-only server letting any caller resume (and thereby execute) a confirmation-gated tool on an UNPUBLISHED roster component. `continue_run`/`cancel_run` now refuse unpublished components whenever the pair exists purely via the ride-along, keyed by (kind, id); with core served or `lifecycle` explicitly included the pair keeps roster-wide REST parity. Verified by mutation and live against the cookbook.
- Trimmed `structuredContent` mirrors the answer text in a `content` key (absorbed from #9765, credited): clients that render structuredContent when present -- Claude Code among them -- previously showed run metadata with no answer, the root cause of the "doesn't work with claude" report.
- Test pins hardened where review showed invocation-only coverage: the ownership gate's refusal must propagate, the HITL loop resolves requirements with the real nested shape, the riding pair's scope gate is pinned, and the exposed-team scope route is asserted.

**Upgrade note**
- A custom tool named like a default tool that will register (or like an earlier custom tool) now fails at `get_app()` with a clear ValueError. On earlier releases FastMCP warned and silently REPLACED, so the colliding config "worked" by shadowing a default tool. Rename the custom tool or scope the default tool out; every other spelling change in this PR remains a silent alias.

**Round-9 closure**
- `as_tool()` markers passed to `Agent(tools=)`/`Team(tools=)` now raise at construction and in the processing chains with a pointer to `MCPConfig.tools` (they previously fell off the classification chain and registered nothing).
- `BaseExternalAgent` adapters joined the kind map (a non-roster adapter can no longer resolve to a same-id roster component of another kind), and `BaseFactory`/`BaseExternalAgent` gained `as_tool()` so every error string prescribing it is truthful for every component family.
- Module rename: `agno/tools/component_tool.py` -> `agno/tools/component.py` (documented import `from agno.tools import ComponentTool` unchanged).

**Security & correctness review (external re-review, all verified by execution)**
- **Cancellation and continuation are bound to the run, not just the named component.** An executed bypass showed that on a default (`user_isolation=False`) server, naming a published component let `cancel_run` cancel a run belonging to an *unpublished* one (cancellation keys on `run_id`, and the ownership verifier no-oped for non-isolated callers). The verifier now proves the run belongs to `(component_type, component_id)` before any tombstone or intent, regardless of user isolation; `session_id` is required to locate the run. `continue_run` gets the same binding. Real-gate regression tests (no stub of the verifier); binding mutation-verified.
- **[P1] `as_tool(name=...)` no longer hides a scope-unsafe id.** The bare exposure path already rejects an out-of-charset id; the override path validated only the model-facing name, so an id like `billing/admin` slipped through — and the RBAC path parser truncates at the slash, letting `agents:billing:run` authorize `billing/admin`. The component id is now validated as a scope-safe segment even under an override. Pinned, mutation-verified.
- **Remote exposures forward the caller's bearer token.** Remote Agent/Team/Workflow run and cancel over MCP dropped the inbound token, so a protected downstream AgentOS 401'd; the token is now forwarded (as the REST routers do), guarded to `BaseRemote`. Mutation-verified. Scope note: this pins bearer *propagation* to `arun`/`acancel_run`; a fully protected downstream is not yet supported end-to-end (the session-writability preflight dereferences the remote's network-backed `db` before the token is used, and scoped/PAT remote cancellation fails closed at the local ownership gate). Remote HITL continuation is likewise deferred. These are functionality gaps, not local authorization holes, and are called out as follow-up work.
- **Wrong-roster placement is refused.** `_locate_component`'s identity fast path returned from whichever roster held the object; a `Team` placed in `agents=` would run under agent scopes/`SessionType.AGENT`. It now refuses a class/roster mismatch.
- **Unreachable remotes no longer stall the build.** Registration skips the network-backed name/description read when `as_tool` supplies both.
- **The paused/no-continue REST-recovery hint is mirrored into `structuredContent`** so structured-only clients (Claude Code) receive it (closes a gap the mirror fix itself opened).

**Cookbook**
- New: `14_mcp/agents_as_tools.py` (the headline shape). Updated to the new spellings: the rest of `14_mcp` + README, `examples/team_brain`, `examples/metrics_desk`, and the `agnoctl` enable-MCP help string.

**Review round 2** (responding to two external reviews)
- Exposed ids are now validated, never sanitized (see above) — resolves the sanitized-name/`continue_run` dead-end and the slash-in-id scope confusion in one rule. The underlying scopes.py behavior (fnmatch `*` spanning `/` while the resource-id capture stops at `/`) pre-exists on main via the generic `run_agent` and is being filed separately.
- The 64-char cap's error text now attributes the limit correctly (strictest LLM-provider limit MCP clients bridge into; the MCP spec itself allows 128 and dots).
- `tools/list` stays a deployer-curated publication list (names/descriptions visible to any caller who can list; invocation stays scope-gated; omit a component to hide it) — now documented on the config field, and the in-code justification corrected (the MCP spec *permits* per-caller lists; per-caller filtering is possible future work, not a spec constraint).
- Custom-tool collision detection now uses the names FastMCP actually registered (catches `functools.partial` / callable-instance names) instead of re-deriving them.
- The deprecated `enable_builtin_tools` alias now covers non-dict Mappings and post-construction assignment; `MCPConfig` is `extra="forbid"` so an `agent=`/`team=` typo fails loudly.
- Wrong-kind exposure (a Team in `MCPConfig.agents`) now says "move it to `MCPConfig.teams`" instead of advising a roster change that would mis-gate scopes.
- agnoctl's enable-MCP help string reverted to `mcp_server=True` (agnoctl versions independently of agno and must print code that works on every released agno) with a note pointing at the new spelling.
- Factory finding from review: MCP run tools (generic *and* exposed) don't carry `factory_input` — a pre-existing MCP-wide parity, now documented on the config; threading `factory_input` through the MCP surface is follow-up work.
- New tests for all of the above, incl. the two-components-of-one-kind late-binding guard, `result_mode="full"` on exposed tools, and the call-time progress label.

**Reshape (round 5, decided with ab):** the earlier `agents=`/`teams=`/`workflows=` params are **dropped** (never released) in favor of the single `tools=` surface above plus `as_tool()`. What carried over unchanged: the whole hardened pipeline (call-time resolution, scopes, session minting, ownership gates, remote-safe metadata, paused-REST hint, name validation, real-registry collisions) and ~90% of the test suite. What's new: the `ComponentTool` marker (declarative — a pre-bound callable would bypass the machinery; and deliberately not named `MCPTool`, which would sit one letter from the client-side `MCPTools`), roster-membership kind derivation replacing the wrong-kind scan machinery, ambiguity errors for cross-roster id copies, and `agent_id`/`team_id`/`workflow_id` in trimmed `structuredContent` (the continue-handle now that names can differ from ids). 68 tests; live-verified (override pitch on the wire, `agent_id` in results, session continuity).

**Review round 4** (Claude round-2 review + human inline comments — with live provider probes)
- **Tool-name rule is now what providers actually accept**, per live probes in review: 128-char cap (OpenAI/Anthropic/Gemini all accept 128, reject 129 — replacing my recalled 64) and a leading letter/underscore (Gemini 400s a leading digit and validates per request, so one bad name failed *every* tool in the reviewer's call).
- The invalid-id error names its source ("auto-derived from `name='Research & Writing Team'`"), suggests a clean candidate via the house id mint + NFKD transliteration (`research-writing-team`, `reviseur`), omits suggestions that would collide, and notes that changing an existing component's id is a migration (sessions/memories are keyed by it).
- **Remote-safe metadata**: `RemoteTeam`/`RemoteWorkflow` `name`/`description` are network-backed properties; exposure reads them through a safe getter so an unreachable remote degrades that tool's description instead of hard-failing `get_app()` (repro'd: previously exposure made boot fail where the non-exposed roster booted fine).
- A paused run on a server without `continue_run` now says so in the result text and points at REST — a runtime hint at the moment of the dead end, instead of a boot warning that can't reliably detect HITL toolsets.
- The `mcp=`/`mcp_server=` conflict error no longer reprs values (component reprs can carry an api_key); assigning a dict to `mcp=` raises `TypeError` instead of enabling the server while silently discarding its settings (`authorize` included).
- Documented: the exposed tool list is fixed at build — components added on a live resync are runnable via the generic run tools immediately, and appear as named tools after restart (deliberate v0 boundary; live tool registration + `listChanged` is the follow-up).
- Verified with a first-party client: Claude Code's MCP client connects to the cookbook server and lists the exposed tools.

**Review round 3**
- Wrong-kind checks now run before the id fallback (by identity across rosters, then by concrete class): AgentOS permits an Agent and a Team to share an id, so a Team entry in `MCPConfig.agents` could previously id-match and silently publish the same-id *Agent* under agent scopes. Repro'd, fixed, and pinned — including the stray-copy variant.
- The `extra="forbid"` test now asserts `extra_forbidden` at `loc=("agent",)` (the broad match could pass via the unrelated zero-tools error).
- Added the collision test that only the real-registry fix can pass: a `functools.partial` custom tool (named "partial" by FastMCP, invisible to the old re-derivation) colliding with an exposed id.

**Out of scope (deliberate)**
- The `stateless` knob and anything touching the MCP 2026-07-28 protocol jump — separate follow-up, per the design doc.
- The pre-existing scopes.py slash fail-open (reachable on main via generic `run_agent`) — separate fix PR.
- Threading `factory_input` through the MCP run tools (generic + exposed) — separate follow-up.
- Docs-site updates (private repo).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: cookbook/05_agent_os/14_mcp/agents_as_tools.py
Comment (kausmeows) [inline cookbook/05_agent_os/14_mcp/agents_as_tools.py:1]: ran this cookbook and connected with claude but does not seem to work-

Image

File: libs/agno/agno/os/app.py
Comment (kausmeows) [inline libs/agno/agno/os/app.py]: But now this reasoning is correct for the generic `run_agent/run_team/run_workflow tools` since they resolve `component_id` at call time. It is not correct for the new exposed components, which are registered as named tools at build time from `MCPConfig.agents/teams/workflows`

for example-

```python
AgentOS(agents=[a], mcp=MCPConfig(agents=[a]))
app = o.get_app()                      # tools/list -> chief
o.agents.append(b); cfg.agents.append(b)
o.resync(app)
o._mcp_app is mcp_app  -> True         # never rebuilt
# a fresh build_mcp_server(o) shows ['chief', 'later']; the mounted app still shows only 'chief'
```

So on a resync-based deployment, a newly added exposed component is silently absent from tools/list until process restart

File: libs/agno/agno/os/app.py [hunk 1]
```diff
@@ -421,15 +425,21 @@ def __init__(
...
-        self.mcp_config: Optional[MCPServerConfig] = None
-        self.mcp_server = mcp_server
+        self.mcp_config: Optional[MCPConfig] = None
+        # ``mcp_server`` is the deprecated alias; both set with different values is a
+        # contradiction, never a guess. ``None`` means unset for both, so an explicit
+        # ``mcp=False`` is distinguishable from the default.
+        if mcp is not None and mcp_server is not None and mcp != mcp_server:
+            raise ValueError(
+                "AgentOS() got both mcp= and its deprecated alias mcp_server= with different "
+                f"values ({mcp!r} vs {mcp_server!r}); pass only mcp=."
```
Comment (harshsinha03) [inline libs/agno/agno/os/app.py]: Reprs the component, so the `api_key` ends up visible in the error.

File: libs/agno/agno/os/mcp.py [hunk 2]
```diff
@@ -746,6 +775,294 @@ async def verify(
...
+
+# An exposed tool is named by its component id VERBATIM -- the id is also the client's
+# handle for continue_run/get_sessions and the resource segment in per-resource scopes
+# (agents:<id>:run), so a name that differs from the id would break HITL resume and make
+# the visible tool name disagree with the scope that grants it. Ids outside this shape
+# are therefore a hard error, never sanitized. The 64-char cap is the strictest limit
+# among the LLM providers MCP clients bridge tool names into (the MCP spec itself allows
+# more); a longer name would be rejected client-side at runtime with no server signal.
+_TOOL_NAME_MAX_LENGTH = 64
+_TOOL_NAME_VALID_RE = re.compile(r"[A-Za-z0-9_-]+")
```
Comment (harshsinha03) [inline libs/agno/agno/os/mcp.py]: **Accepts a leading digit, which Gemini rejects.** `Agent(name="2024 Reporter")` with no explicit `id=` mints `2024-reporter`, passes this regex, and gets published:

```
agno serves:  tools published = ['2024-reporter']
real gemini-2.5-flash client over MCP:
  400 Bad Request — function_declarations[0].name: Invalid function name.
                    Must start with a letter or an underscore.
same server, OpenAI client: works
```

Gemini validates `function_declarations` per request, so one bad name fails the whole call and every other exposed tool goes down with it. `[A-Za-z_][A-Za-z0-9_-]*` covers it.

**The 64 cap is stricter than every provider named.** Real calls:

```
OpenAI Responses   128 accepted   129 rejected
Anthropic          128 accepted   129 rejected
Gemini             128 accepted   129 rejected
```

The MCP SDK's own validator warns on anything outside `^[A-Za-z0-9._-]{1,128}$` (advisory — it logs, it doesn't reject). A 65-char id is a hard build failure today for no runtime reason.

**The suggestion at :806 could reuse `generate_component_id_from_name`**, already the mint for `/components` and `StudioTools`:

```
id                 this suggests      helper gives
ops-&-risk         ops---risk         ops-risk
chief-(v2)         chief--v2          chief-v2
```

It won't cover everything, though — it preserves unicode by design (so `validate_component_id` and it agree) and doesn't touch a leading digit, so `réviseur`, `研究员` and `2024-reporter` still need their own branch on top of it.

**A non-Latin id gets no advice at all**, because `candidate` comes back empty and the sentence is dropped:

```
Agent(name="Réviseur") -> id 'réviseur'
  ... outside [A-Za-z0-9_-]. For example, set id='r-viseur' on the component.
                                                 ^^^^^^^^^ é deleted, not transliterated
Agent(name="研究员")    -> id '研究员'
  ... outside [A-Za-z0-9_-].
                            ^ ends here, no instruction
```

The length branch below always ends "Set a shorter id on the component."; this one should too.

File: libs/agno/agno/os/mcp.py [hunk 3]
```diff
@@ -746,6 +775,294 @@ async def verify(
...
+    custom_tool_names: "Optional[Dict[str, str]]" = None,
+) -> None:
+    """Register the components named in ``MCPConfig.agents/teams/workflows`` as tools.
+
+    Names are component ids verbatim (validated, never sanitized -- the id is the
+    handle for continue_run and the per-resource scope segment, so the visible tool
+    name must equal it); collisions with the enabled default tools, custom tools, or
+    other exposed components are a hard error at build.
+
+    The registered tool list is deliberately static and identical for every caller:
```
Comment (harshsinha03) [inline libs/agno/agno/os/mcp.py:1315]: Registration can't filter — there's no caller at build. But `tools/list` has one, and `get_agentos_config` already filters the same roster there. Worth `on_list_tools` if the two should agree.

File: libs/agno/agno/os/mcp.py [hunk 4]
```diff
@@ -746,6 +775,294 @@ async def verify(
...
+        for other_kind in ("agents", "teams", "workflows"):
+            if other_kind == kind:
+                continue
+            for component in getattr(os, other_kind, None) or []:
+                if getattr(component, "id", None) == entry:
+                    raise ValueError(
+                        f"MCPConfig.{kind} contains id {entry!r}, which is one of this AgentOS's "
+                        f"{other_kind}; move it to MCPConfig.{other_kind}."
+                    )
+        raise ValueError(
```
Comment (harshsinha03) [inline libs/agno/agno/os/mcp.py]: Should components from the catalog be included here too? `run_agent` already runs them.

File: libs/agnoctl/agnoctl/discovery.py [hunk 5]
```diff
@@ -391,10 +391,13 @@ def discover_all(
...
 
+# Deliberately the OLD spelling: agnoctl versions independently of agno and talks to
+# servers on any agno release. ``mcp=`` only exists on newer agno; ``mcp_server=`` is a
+# kept alias there, so this line works everywhere. Flip once older agno is out of support.
 MCP_ENABLE_INSTRUCTIONS = """MCP is not enabled on this AgentOS. Enable it and restart:
 
     agent_os = AgentOS(
         agents=[...],
-        mcp_server=True,  # add this line
+        mcp_server=True,  # add this line (spelled mcp= on newer agno)
```
Comment (harshsinha03) [inline libs/agnoctl/agnoctl/discovery.py:401]: worth naming the version here

File: libs/agno/agno/os/config.py [hunk 6]
```diff
@@ -15,26 +15,62 @@
...
+    # invoked over MCP at all (neither via run_agent nor via exposure -- the MCP run
+    # tools do not carry factory_input yet); invoke those over REST.
+    #
+    # The generated tools run through the same machinery as ``run_agent`` /
+    # ``run_team`` / ``run_workflow`` (per-run copies, scope checks, session minting,
+    # ownership gates, progress reporting), so RBAC scopes like ``agents:run`` apply
+    # to them identically. Combine with ``default_tools=False`` to serve ONLY the
+    # named components.
+    #
+    # Note for HITL: resuming a PAUSED run needs the default ``continue_run`` tool, so
```
Comment (harshsinha03) [inline libs/agno/agno/os/config.py]: Worth a warning or debug here for the case where a `requires_confirmation` tool pauses and there is no `continue_run` registered?

File: cookbook/05_agent_os/14_mcp/agents_as_tools.py
Comment (ashpreetbedi) [reply cookbook/05_agent_os/14_mcp/agents_as_tools.py:1]: Verified with a first-party client at HEAD: Claude Code's MCP client connects to this cookbook server over streamable HTTP (`claude mcp add --transport http t http://localhost:7777/mcp` → ✔ Connected) and lists `chief`/`researcher`. Which Claude surface did you use — claude.ai web, Claude Desktop, or Claude Code — and was the server reached through a tunnel/public URL? If it was tunneled: an open (no-auth) server deliberately defaults to localhost-only Host validation (DNS-rebinding protection), so a tunnel host gets 400 until you pass `MCPConfig(allowed_hosts=["your-host"])`. claude.ai web additionally can't reach localhost at all and needs `mcp_auth` (see `oauth_builtin.py`). If you saw something else, the screenshot text would help me reproduce.

File: libs/agno/agno/os/app.py
Comment (ashpreetbedi) [reply libs/agno/agno/os/app.py]: Confirmed — your repro is accurate, and it's a deliberate v0 boundary rather than an oversight: the mounted MCP app can't be rebuilt after startup (the streamable-HTTP session manager is started by the app lifespan; rebuilding orphans it — same reason `_mcp_app` is cached), and the spec scopes dynamic exposure out of v0. The mitigation that exists today: components added on resync are runnable through the generic `run_agent`/`run_team`/`run_workflow` immediately (they resolve at call time) — only the *named* tool list is fixed at build. As of 7b7c07f3e that's documented on the config field and in the README. The real fix is registering new tools on the live FastMCP instance at resync (`mcp.add_tool` works post-build) + collision revalidation + `listChanged` — follow-up shape, same thread as catalog exposure.

File: libs/agno/agno/os/app.py (hunk 1 above)
Comment (ashpreetbedi) [reply libs/agno/agno/os/app.py]: Fixed in 7b7c07f3e — the message no longer reprs the values at all (an MCPConfig can hold component instances whose repr may carry a model api_key).

File: libs/agno/agno/os/mcp.py (hunk 2 above)
Comment (ashpreetbedi) [reply libs/agno/agno/os/mcp.py]: Adopted wholesale in 7b7c07f3e — thank you for probing this against real providers, it overturned two of my positions: regex is now `[A-Za-z_][A-Za-z0-9_-]*` (your Gemini leading-digit case, with the per-request blast radius noted in the comment), the cap is 128 (your 128-accepted/129-rejected probes beat my recalled 64), the suggestion routes through `generate_component_id_from_name` plus an NFKD/ASCII fold first (so `réviseur` suggests `reviseur`, not `r-viseur`), and the non-Latin/empty-candidate branch now always ends with an instruction, mirroring the length branch. All covered by tests, including `2024-reporter` → suggests `agent-2024-reporter`.

File: libs/agno/agno/os/mcp.py (hunk 3 above)
Comment (ashpreetbedi) [reply libs/agno/agno/os/mcp.py:1315]: Agreed the mechanism exists — the spec permits per-caller lists (SEP-2549 `cacheScope` exists precisely for that), and `on_list_tools` is the right hook. Kept unfiltered deliberately for v0: exposure is a hand-typed publication list (unlike `get_agentos_config`, which enumerates the whole deployment — hence its filter), custom tools already behave this way, and omitting a component from the config hides it while keeping it runnable via `run_agent`. The trade-off is now documented on the config field. Happy to take scope-filtered `tools/list` via `on_list_tools` as a follow-up PR if we want the two surfaces to agree.

File: libs/agno/agno/os/mcp.py (hunk 4 above)
Comment (ashpreetbedi) [reply libs/agno/agno/os/mcp.py]: Deliberate v0 non-goal (per the design doc): catalog/DB-registry components would make the named tool list drift at runtime, which needs the tool-list refresh story (live `add_tool` + collision revalidation + `listChanged`) — same thread as the resync staleness comment above. `run_agent` covering them in the meantime is exactly the intended split: the generic tools are the dynamic surface, exposure is the static published one.

File: libs/agnoctl/agnoctl/discovery.py (hunk 5 above)
Comment (ashpreetbedi) [reply libs/agnoctl/agnoctl/discovery.py:401]: Deliberately vague for now: agnoctl versions independently of agno and can't introspect the server's agno version, and the release that carries `mcp=` doesn't exist yet — naming a version here before it's cut risks printing a wrong floor. Will pin the exact version in this string when the carrying release is tagged.

File: libs/agno/agno/os/config.py (hunk 6 above)
Comment (ashpreetbedi) [reply libs/agno/agno/os/config.py]: Went with a runtime hint instead of a boot warning: build-time detection is unreliable (tools can be attached dynamically or via factories), and a warning would fire on configs that never pause. As of 7b7c07f3e, a paused run on a server without `continue_run` says so in the tool result itself — "The continue_run tool is not registered on this server; resume this run over the REST API, or serve the core default tools alongside the exposed components." — which lands exactly at the moment the dead end happens. Tested both directions (hint present with `default_tools=False`, absent when `core` tools are served).

**Comment by ashpreetbedi** [comment]
Heads-up for reviewers: the exposure API reshaped at a0891a32d (decided with @ashpreetbedi). The `agents=`/`teams=`/`workflows=` params are gone; `MCPConfig.tools` is the one surface — bare components expose with derived defaults, and `component.as_tool(name=..., description=...)` publishes under an explicit model-facing name/pitch (declarative marker, so the hardened run pipeline — scopes, session minting, ownership, HITL, progress — still applies; a pre-bound callable would have bypassed it).

Everything from the earlier review rounds carried over: provider-probed name validation (now also applied to `as_tool` overrides), real-registry collision detection, remote-safe metadata, the paused-REST hint, and the roster rule — with kind now derived from roster membership, which replaces the wrong-kind scan machinery and turns the cross-roster same-id case into an explicit ambiguity error. New since the tool name can differ from the id: trimmed `structuredContent` carries `agent_id`/`team_id`/`workflow_id` as the `continue_run`/`get_sessions` handle. The PR description is current; fresh eyes on `_split_tool_entries`/`_locate_component` and the as_tool tests would be most valuable.

File: libs/agno/agno/os/app.py
Comment (ashpreetbedi) [reply libs/agno/agno/os/app.py]: Confirmed — ran your exact sequence (the surface is `MCPConfig(tools=[...])` post-reshape, same result): resync keeps the mounted app, a fresh `build_mcp_server(os)` shows the new tool, so a resync-added exposure is absent from `tools/list` until restart.

Two-part disposition:

1. The comment you quoted predates exposures and now overclaims. It stays true for the builtin run tools (call-time id resolution — a resync-added component is immediately runnable through `run_agent`), but exposure *registration* is a build-time act on the FastMCP server. Fixing the comment in this PR, and sharpening the README line while at it: under `default_tools=False` the new component isn't reachable over MCP at all until restart, not merely unnamed.

2. The rebuild in your snippet can't be the runtime fix — a freshly built app mounted mid-flight has a StreamableHTTP lifespan that never runs, so every `/mcp` request 500s. The viable path is retaining the FastMCP server object (we currently keep only the wrapped http_app) and `add_tool`-ing the exposure diff on the live server during resync — FastMCP consults its tool manager per request, so new tools appear without touching the transport. But that pulls in runtime collision policy, the lifecycle ride-along flipping on when the first exposure appears, removal semantics, and `tools/list_changed` notifications — carving it out as a follow-up rather than growing it in this round. Worth noting the blast radius is narrow: resync's real users (Studio, the lifespan cookbook) add to the *roster*, which the generic run tools pick up live; exposure requires hand-mutating `MCPConfig.tools` and resyncing.

Good catch on the comment drift.

File: cookbook/05_agent_os/14_mcp/agents_as_tools.py
Comment (ashpreetbedi) [reply cookbook/05_agent_os/14_mcp/agents_as_tools.py:1]: Root-caused — your screenshot was the key, and it was never a connectivity problem. Claude Code renders `structuredContent` when a tool result carries one, and ours carried only `{run_id, session_id, status}` — the answer lived in the `content` blocks Claude Code doesn't show in that case. So the run genuinely completed ("COMPLETED" in your screenshot) and the answer was on the wire, just not in the field your client displays.

Fixed in `2e406d90f`: trimmed `structuredContent` now mirrors the answer text in a `content` key, the same pattern the official MCP servers use (the run tools declare no outputSchema, so nothing constrains the added key). This is the fix from #9765 — absorbed here with credit to Mustafa, since this cookbook is the surface that hits it and both PRs edit the same function; the toolkit-flattening and schema work stay in #9765.

Verified live against this cookbook: `structuredContent` keys are now `agent_id/content/run_id/session_id/status` with `content == "Earth"` matching the text block. Re-running your Claude Code session against the branch should now show the answer.
